crates/sp-subspace-mmr/src/host_functions.rs

```diff
@@ -0,0 +1,58 @@
+use crate::runtime_interface::LeafData;
+use sp_blockchain::HeaderBackend;
+use sp_core::H256;
+use sp_runtime::traits::{Block as BlockT, Header as HeaderT};
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+/// Trait to query MMR specific data through host function..
+pub trait SubspaceMmrHostFunctions: Send + Sync {
+    /// Returns the MMR Leaf data for given consensus block hash
+    fn get_mmr_leaf_data(&self, consensus_block_hash: H256) -> Option<LeafData>;
+}
+
+sp_externalities::decl_extension! {
+    pub struct SubspaceMmrExtension(Arc<dyn SubspaceMmrHostFunctions>);
+}
+
+impl SubspaceMmrExtension {
+    /// Create a new instance of [`SubspaceMmrExtension`].
+    pub fn new(inner: Arc<dyn SubspaceMmrHostFunctions>) -> Self {
+        Self(inner)
+    }
+}
+
+/// Implementation of MMR host function.
+pub struct SubspaceMmrHostFunctionsImpl<Block, Client> {
+    consensus_client: Arc<Client>,
+    _phantom: PhantomData<Block>,
+}
+
+impl<Block, Client> SubspaceMmrHostFunctionsImpl<Block, Client> {
+    pub fn new(consensus_client: Arc<Client>) -> Self {
+        SubspaceMmrHostFunctionsImpl {
+            consensus_client,
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<Block, Client> SubspaceMmrHostFunctions for SubspaceMmrHostFunctionsImpl<Block, Client>
+where
+    Block: BlockT,
+    Block::Hash: From<H256>,
+    Client: HeaderBackend<Block>,
+{
+    fn get_mmr_leaf_data(&self, consensus_block_hash: H256) -> Option<LeafData> {
+        let header = self
+            .consensus_client
+            .header(consensus_block_hash.into())
+            .ok()
```

Comment:
I'm not sure `.ok()` is the right pick here. There is a big difference between block not existing and database error. Since this is a host function, it should probably panic instead of silently returning `None` to avoid inconsistencies between nodes on the network.

Reply:
Should panic through `expect`. Let me know if you have a better message there :)

Reply:
This is not what I meant. There is nothing wrong with block not existing, caller can do `.expect()` is they know block exists for sure because it is a parent block. I only talked about `.ok()` being replaced with `.expect()`.

`.expect()` message should have a proof, not a hope. So in this case it'll be something like "Database error is fatal in host function, there is no recovery from this; qed".

Reply:
While I agree that outer result should not happen, the host should not anyway give the optional result leading to forks on the network. The change was to solve that

Reply:
Optional result here is the correct API (function can't guarantee which block hash it is called with), but if called from within runtime with parent block hash, then you need to use `.expect()` at the call site since there you do know whether block hash must exist or not. As such, you still get your data unconditionally and don't have any more forks than with what you proposed.